activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -246,6 +258,14 @@ def key_conversion_required?
             @key_conversion_required
           end
 
+          def derive_key(owner, key)
+            if key.is_a?(Array)
```

Comment:
Eventually we should be able to get rid of the branching and make key always an array even if it's a single-value array like `[:id]`. However currently the dual nature of `key` allows us to determine when we are going to query models associated by a composite key in `load_records_for_keys`